### ISSUE TITLE
Conch relies on a CPython implementation detail

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5779]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#5779|
|Type|enhancement|
|Created|2012-07-14 18:03:47Z|
|Branch|https://github.com/twisted/twisted/tree/conch-should-not-use-builtins-5779|

It uses `__builtins__` and assumes that it will be a dict.  According to http://docs.python.org/library/__builtin__.html this is a) an implementation detail of CPython, and b) not guaranteed to be a dict! it will sometimes be a module.

Attachments:

* [t5779-builtins.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/345/3459fd8f5bb6e4b27e921451f173c15d0f36ef23/d7d8760add00485483480eab390fc369153d8409.diff) (575 bytes) - added by Alex on 2012-07-14 18:04:12Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5779 5779
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__branches_conch_should_not_use_builtins_5779 branches/conch-should-not-use-builtins-5779
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1342289027000000 1342289027000000
changetime__1342344219000000 1342344219000000
version__None None
owner__ 
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [34803]) Branching to 'conch-should-not-use-builtins-5779'

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|spiv commented|
|-|-|

I suppose strictly speaking this is correct, but:

 * is gmpy even available for non-CPython interpreters, and
 * if it is available, are we sure it's faster?

e.g. I'm wondering if for pypy we'd be better off not doing this hack.  I guess optimising exponentiation isn't pypy's focus, but who knows…

Anyway, this patch does seem to be slightly more correct in theory, so we may as well use it :)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [34804]) Apply Alex's t5779-builtins.diff patch refs [#5779](https://github.com/twisted/twisted/issues/5779)

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|Alex commented|
|-|-|

spiv: No, it's not, and in fact from reading the source I seriously doubt it will ever compile against either CPyExt (PyPy's compatibility layer) or IronClad (IronPython).  This was submitted purely as a correctness issue, not a compatibility one.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [34807]) Add a newsfile refs [#5779](https://github.com/twisted/twisted/issues/5779)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

Thanks for finding and fixing this Alex!  In the future, could you also include a news file in your patch?  http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

[buildbot tests pass](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-should-not-use-builtins-5779).  Merging this.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli set status to `closed`|
|-|-|

(In [34811]) Merge conch-should-not-use-builtins-5779: conch should not depend on
__builtlins__ being a dict

Author: Alex
Reviewer: cyli, spiv
Fixes: [#5779](https://github.com/twisted/twisted/issues/5779)

Conch uses __builtins__ and assumes that it will be a dict, which is an
implementation of CPython and not guaranteed to dict instead of a module

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

Oops, forgot to remove review keyword
